packages/openai: New OpenAI integration (#12494)
elastic/integrations: packages/openai/_dev/build/docs/README.md

# OpenAI

The OpenAI integration allows you to monitor OpenAI API usage metrics. OpenAI is an AI research and deployment company that offers [API platform](https://openai.com/api) for their industry-leading foundation models.

With the OpenAI integration, you can track API usage metrics across their models, as well as for vector store and code interpreter. You will use Kibana to visualize your data, create alerts if usage limits are approaching, and view metrics when you troubleshoot issues. For example, you can track token usage and API calls per model.

## Data collection

The OpenAI integration leverages the [OpenAI Usage API](https://platform.openai.com/docs/api-reference/usage) to collect detailed usage metrics. The Usage API delivers comprehensive insights into your API activity, helping you understand and optimize your organization's OpenAI API usage.

## Data streams

> The OpenAI integration collects the following logs data streams:

Reviewer: It is mentioned - "logs data stream". Shouldn't this be metrics?

# OpenAI

> The OpenAI integration allows you to monitor OpenAI API usage and performance. OpenAI is a leading AI platform providing various AI models and APIs for natural language processing, image generation, and other AI capabilities.

Reviewer 2: Question: should we say "foundation models" instead of "AI models"?

Author: Yes, I updated it. Now it's very close to OpenAI's own website and very crisp.

Ref: 
- https://openai.com/about/
- https://openai.com/api
- etc.

With the OpenAI integration, you can track API usage, costs, and performance metrics across your OpenAI implementations. You will use Kibana to visualize your data, create alerts if usage limits are approaching, and view metrics when you troubleshoot issues.

The OpenAI integration lets you monitor your metrics across the different OpenAI models. For example, you can track token usage and API calls per model.

## Data collection

The OpenAI integration collects comprehensive usage data through the [OpenAI Usage API](https://platform.openai.com/docs/api-reference/usage). The Usage API provides detailed insights into your activity across the OpenAI API and helps you understand how your organization is using OpenAI APIs.

> 

Reviewer 2: are we only tracking how your org is using the APIs (i.e. usage) or are we also giving you an understanding of how those OpenAI models are performing in terms of latency, errors,  etc?

Author: Yes, only usage. Latency, errors, etc. are not present.

See: https://github.com/elastic/observability-dev/issues/3904#issuecomment-2469711313 for more detailed explanation.

## Data streams

The OpenAI integration collects the following logs data streams:

- `audio_speeches`: Collects audio speeches usage events for the organization.
- `audio_transcriptions`: Collects audio transcriptions usage events for the organization.
- `code_interpreter_sessions`: Collects code interpreter sessions usage events for the organization.
- `completions`: Collects completions usage events for the organization.
- `embeddings`: Collects embeddings usage events for the organization.
- `images`: Collects images usage events for the organization.
- `moderations`: Collects moderations usage events for the organization.
- `vector_stores`: Collects vector stores usage events for the organization.

See more details for data streams in the [Logs](#logs-reference).

## Requirements

> You need Elasticsearch for storing and searching your data and Kibana for visualizing and managing it.

Reviewer 2: Please work with @alaudazzi  to make sure the requirements section presence (or absence) is aligned with all other  integration pages

Author: I did take a look at many "Requirements" sections; looks similar.

Reviewer 3: LGTM

You need an OpenAI account to access the [Usage API](https://platform.openai.com/docs/api-reference/usage) with a valid [Admin key](https://platform.openai.com/settings/organization/admin-keys) to programmatic access to Usage APIs.

> 

Reviewer 2: There seems to be some unnecessary repetition in this sentence, can we simplify:
"You need an OpenAI account to access the Usage API with a valid Admin key to programmatic access to Usage APIs."
@alaudazzi  can you help please?

Author: Based on what I've seen in other READMEs:

- The Requirements section should only list the requirements, nothing more.
- The Setup section elaborates on how to perform the actual setup. Also, made "### Generate an Admin key" a subsection under "## Setup"

So, in Requirements I listed what's required, and in the Setup section, I explained how to generate one Admin Key.

Should I club them?

Reviewer 3: @shmsr these two sections -- Requirements and Setup -- should remain separate

## Setup

For step-by-step instructions on how to set up an integration, see the [Getting started](https://www.elastic.co/guide/en/starting-with-the-elasticsearch-platform-and-its-solutions/current/getting-started-observability.html) guide.

### Generate an Admin key

> To generate an Admin key, please generate a key or use an existing one from the [Admin keys](https://platform.openai.com/settings/organization/admin-keys) page. Use the Admin key to configure the OpenAI integration.

Reviewer 2: Shouldn't this section be right next to the section that talks about needing an OpenAI account and Admin keys? We should decide of it will be before or after that section given might be a dependency. @alaudazzi  any thoughts on this?

Author: https://github.com/elastic/integrations/pull/12494#discussion_r1938502879

## Collection behavior

> By default, the OpenAI integration fetches events with a bucket width of 1 day (`1d`), which means events are aggregated by day. Events are collected from the set initial start time until the current time except the bucket that is currently being used by OpenAI, that is the last bucket is skipped as it is not complete. So, based on configured bucket width, the integration collects events from the initial start time until the current time minus the bucket width.

Reviewer 2: is "bucket width" a standard term we use? Will customers understand what we mean? What do we say in other integrations and can we use the same language if it's different?

Author: So, we do give flexibility to the users to change the "bucket_width". It is not our terminology but OpenAI's. Check the 3rd query parameter here: https://platform.openai.com/docs/api-reference/usage/completions

So, by default, we keep this as `1d` (one day). Infact, OpenAI's own dashboard works on daily aggregated data as well; no knobs there. But as their new API gives the flexibility to change it, that's why we have made it configurable.

So, concept of bucket I explained here because it is important to make the user understand how the collection works:

So, let's think about last 2 days (Feb 2, Feb 3). The bucket_width is set to `1d`, that is OpenAI will return aggregated data for a day. So, each bucket has a start and end timestamp i.e., for Feb 2, you have one bucket, and for Feb 3, another bucket.

Now, as today's Feb 3, the ended hasn't ended yet. So, if you collect data, as the bucket is incomplete as still filling, we won't be able to get the correct data. Like we can collect, but the aggregation on our side, and the visualizations will not match up with OpenAI's. So, incomplete buckets are left out. As soon as the bucket is closed and it is not being filled anymore, we collect it.

## Logs reference

**ECS Field Reference**

Refer to this [document](https://www.elastic.co/guide/en/ecs/current/ecs-field-reference.html) for detailed information on ECS fields.

### Audio speeches

> The `audio_speeches` data stream captures audio speeches usage events for the organization.

Reviewer 2: Do we need to repeat "for the organization" everywhere? Isn't it assumed?  @alaudazzi  would it make sense to get rid of this?

Author: Removed this. Makes sense to not repeat it.

# OpenAI

The OpenAI integration allows you to monitor OpenAI API usage and performance. OpenAI is a leading AI platform providing various AI models and APIs for natural language processing, image generation, and other AI capabilities.

> Use the OpenAI integration to track API usage, costs, and performance metrics across your OpenAI implementations. Then visualize that data in Kibana, create alerts to notify you if usage limits are approaching, and view metrics when troubleshooting issues.

Reviewer 3: ```suggestion
With the OpenAI integration, you can track API usage, costs, and performance metrics across your OpenAI implementations. You will use Kibana to visualize your data, create alerts if usage limits are approaching, and view metrics when you troubleshoot issues.
```

> The integrations lets you see your metrics split across the different OpenAI models. For example, you can track token usage and API calls per model.
> 

Reviewer 3: ```suggestion
The OpenAI integration lets you monitor your metrics across the different OpenAI models. For example, you can track token usage and API calls per model.

```

## Data collection

The OpenAI integration collects comprehensive usage data through the [OpenAI Usage API](https://platform.openai.com/docs/api-reference/usage). The Usage API provides detailed insights into your activity across the OpenAI API and helps you understand how your organization is using OpenAI APIs.

## Data streams

The OpenAI integration collects the following logs data streams:

Data streams:

> 

Reviewer 3: Remove this line

- `audio_speeches`: Collects audio speeches usage events for the organization.
- `audio_transcriptions`: Collects audio transcriptions usage events for the organization.
- `code_interpreter_sessions`: Collects code interpreter sessions usage events for the organization.
- `completions`: Collects completions usage events for the organization.
- `embeddings`: Collects embeddings usage events for the organization.
- `images`: Collects images usage events for the organization.
- `moderations`: Collects moderations usage events for the organization.
- `vector_stores`: Collects vector stores usage events for the organization.

See more details for data streams in the [Logs](#logs-reference).

## Requirements

You need Elasticsearch for storing and searching your data and Kibana for visualizing and managing it.

You need an OpenAI account to access the [Usage API](https://platform.openai.com/docs/api-reference/usage) with a valid [Admin key](https://platform.openai.com/settings/organization/admin-keys). Admin keys are being used for programmatic access to endpoints detailed in the [Usage API](https://platform.openai.com/docs/api-reference/usage).

> 

Reviewer 3: Do we need a Setup section here?

Author: Added.

### Collection behavior

By default, the OpenAI integration fetches events with a bucket width of 1 day (`1d`), which means events are aggregated by day. Events are collected from the set initial start time until the current time except the bucket that is currently being used by OpenAI, that is the last bucket is skipped as it is not complete. So, based on configured bucket width, the integration collects events from the initial start time until the current time minus the bucket width.

## Logs reference

### Audio speeches

The `audio_speeches` data stream captures audio speeches usage events for the organization.

{{event "audio_speeches"}}

**ECS Field Reference**

> Please refer to the following [document](https://www.elastic.co/guide/en/ecs/current/ecs-field-reference.html) for detailed information on ECS fields.

Reviewer 3: ```suggestion
Refer to this [document](https://www.elastic.co/guide/en/ecs/current/ecs-field-reference.html) for detailed information on ECS fields.
```

{{fields "audio_speeches"}}

### Audio transcriptions

The `audio_transcriptions` data stream captures audio transcriptions usage events for the organization.

{{event "audio_transcriptions"}}

**ECS Field Reference**

> Please refer to the following [document](https://www.elastic.co/guide/en/ecs/current/ecs-field-reference.html) for detailed information on ECS fields.

Reviewer 3: This paragraph is repeated 7 times. It is probably enough to keep it only once, at the top of the section?

# OpenAI

The OpenAI integration allows you to monitor OpenAI API usage and performance. OpenAI is a leading AI platform providing various AI models and APIs for natural language processing, image generation, and other AI capabilities.

Use the OpenAI integration to track API usage, costs, and performance metrics across your OpenAI implementations. Then visualize that data in Kibana, create alerts to notify you if usage limits are approaching, and view metrics when troubleshooting API issues.

> For example, if you wanted to monitor costs across different OpenAI models, you could track token usage and API calls per model. Then you can visualize various trends in Kibana dashboards, set up alerts for unusual usage spikes, or troubleshoot by analyzing the number of requests.

Reviewer 2: Change "For example, if you wanted to monitor costs across different OpenAI models, you could track token usage and API calls per model. Then you can visualize various trends in Kibana dashboards, set up alerts for unusual usage spikes, or troubleshoot by analyzing the number of requests."
TO:
"The integrations lets you see your metrics split across the different OpenAI models. For example, you can track token usage and API calls per model."
